## Imports 

In [9]:
import pandas as pd
import numpy as np
from ydata_profiling import ProfileReport

df = pd.read_csv("data.csv")
profile = ProfileReport(df, title="My Data Report")

# Save as HTML file to view in browser
profile.to_file("data_report.html")
print("Report saved as data_report.html - open it in your browser")


Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]

100%|██████████| 14/14 [00:00<00:00, 238.71it/s]


Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

Report saved as data_report.html - open it in your browser


In [10]:
df.dtypes

age             int64
sex             int64
cp              int64
trestbps       object
chol           object
fbs            object
restecg        object
thalach        object
exang          object
oldpeak       float64
slope          object
ca             object
thal           object
num             int64
dtype: object

In [11]:
df.drop_duplicates(inplace=True)
df.duplicated().sum()




np.int64(0)

In [12]:
# Check column names and shape
print("Columns in dataframe:")
print(df.columns.tolist())
print(f"\nShape: {df.shape}")
print(f"\nFirst few rows:")
df.head()


Columns in dataframe:
['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach', 'exang', 'oldpeak', 'slope', 'ca', 'thal', 'num       ']

Shape: (293, 14)

First few rows:


age  sex  cp trestbps chol fbs restecg thalach exang  oldpeak slope ca  \
0   28    1   2      130  132   0       2     185     0      0.0     ?  ?   
1   29    1   2      120  243   0       0     160     0      0.0     ?  ?   
2   29    1   2      140    ?   0       0     170     0      0.0     ?  ?   
3   30    0   1      170  237   0       1     170     0      0.0     ?  ?   
4   31    0   2      100  219   0       1     150     0      0.0     ?  ?   

  thal  num         
0    ?           0  
1    ?           0  
2    ?           0  
3    6           0  
4    ?           0

In [13]:
# Clean column names first
df.columns = df.columns.str.strip()

# Replace '?' with NaN across entire dataframe
df = df.replace('?', 0)

# Convert numeric columns to proper numeric type
numeric_cols = ['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 
                'thalach', 'exang', 'oldpeak', 'slope', 'ca', 'thal']

for col in numeric_cols:
    df[col] = pd.to_numeric(df[col], errors='coerce')

# Fill missing values with median or 0
df = df.fillna(0)

target_name = "num"

x = df.drop(columns="num")
y = df[target_name]
x.head()


age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   28    1   2       130   132    0        2      185      0      0.0      0   
1   29    1   2       120   243    0        0      160      0      0.0      0   
2   29    1   2       140     0    0        0      170      0      0.0      0   
3   30    0   1       170   237    0        1      170      0      0.0      0   
4   31    0   2       100   219    0        1      150      0      0.0      0   

   ca  thal  
0   0     0  
1   0     0  
2   0     0  
3   0     6  
4   0     0

In [14]:
df["oldpeak"].value_counts()

oldpeak
0.0    188
1.0     41
2.0     31
1.5     16
3.0      9
2.5      3
0.5      2
0.8      1
4.0      1
5.0      1
Name: count, dtype: int64

In [15]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=42)

In [19]:
# import numpy as np
# from sklearn.linear_model import LogisticRegression
# from sklearn.metrics import f1_score, precision_score, recall_score, accuracy_score
# from sklearn.model_selection import GridSearchCV

# # Define the model
# reg = LogisticRegression(max_iter=1000)

# # Define the parameter grid
# param_grid = {
#     'C': [0.01, 0.1, 1, 10, 100],          # Regularization strength
#     'penalty': ['l1', 'l2', 'elasticnet', 'none'],  # Type of regularization
#     'solver': ['liblinear', 'saga']        # Solvers that support l1/elasticnet
# }

# # Apply GridSearchCV
# grid_search = GridSearchCV(
#     estimator=reg,
#     param_grid=param_grid,
#     cv=5,                # 5-fold cross-validation
#     scoring='accuracy',  # You can also use 'f1', 'precision', 'recall'
#     n_jobs=-1            # Use all cores for faster computation
# )

# # Fit the model
# grid_search.fit(x_train, y_train)

# # Best parameters
# print("Best Parameters:", grid_search.best_params_)

# # Best estimator
# best_reg = grid_search.best_estimator_

# # Predictions with best model
# lr_pred = best_reg.predict(x_test)

# # Evaluation metrics
# print("accuracy_score:", accuracy_score(y_test, lr_pred))
# print("\nF1:", f1_score(y_test, lr_pred))
# print("\nPrecision score:", precision_score(y_test, lr_pred))
# print("\nRecall score:", recall_score(y_test, lr_pred))


In [17]:
import numpy as np
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import  f1_score, precision_score, recall_score, accuracy_score, confusion_matrix
reg = LogisticRegression(max_iter= 10000)
reg.fit(x_train,y_train)

lr_pred=reg.predict(x_test)

print("accuracy_score:",accuracy_score(y_test,lr_pred))
print("\n F1:\n",f1_score(y_test,lr_pred))
print("\n Precision score is:\n",precision_score(y_test,lr_pred))
print("\n Recall score is:\n",recall_score(y_test,lr_pred))
print( "confusion matrix is:\n",confusion_matrix(y_test,lr_pred))

accuracy_score: 0.8135593220338984

 F1:
 0.7441860465116279

 Precision score is:
 0.64

 Recall score is:
 0.8888888888888888
confusion matrix is:
 [[32  9]
 [ 2 16]]


# Model Performance Evaluation: Logistic Regression

Following the data cleaning and preprocessing steps, a Logistic Regression model (max_iter=10000) was trained and tested. Below are the finalized results.

### 1. Key Performance Metrics
| Metric | Score |
| :--- | :--- |
| **Accuracy Score** | **81.36%** |
| **Recall Score** | **88.89%** |
| **F1 Score** | **74.42%** |
| **Precision Score** | **64.00%** |

### 2. Confusion Matrix Analysis
The confusion matrix provides a detailed look at where the model is succeeding and where it is making errors:

| | Predicted: No Disease | Predicted: Disease |
| :--- | :---: | :---: |
| **Actual: No Disease** | 32 (True Negative) | 9 (False Positive) |
| **Actual: Disease** | 2 (False Negative) | 16 (True Positive) |


### 3. Final Observations
* **High Sensitivity:** The model demonstrates a very high **Recall (88.89%)**, which is critical in medical diagnostics. It only missed 2 actual cases of heart disease (False Negatives).
* **Reliability:** With an **Accuracy of 81.36%**, the model is a strong baseline for this dataset.
* **Room for Improvement:** The **Precision (64.00%)** indicates that there are several "False Alarms" (9 cases). Future work could focus on reducing these False Positives to avoid unnecessary follow-up testing for healthy patients.
